### DISCUSSION TITLE
Motivation for developing this simulator and project goals

### DISCUSSION THREAD

**Discussion by rgov** [osrf/mbari_wec#60] [General]
Congrats to the MBARI and OSRF teams for achieving the milestone of the initial Gazebo simulation release, and thanks for generously sharing it as open source.

The project website says,

> This project provides a software interface to the system to allow [control] algorithms to be efficiently developed, tested, and executed. 

Could the team share more details on the motivation behind developing this simulator? Is the goal to make it possible to optimize these "control algorithms" to maximize energy harvested, or to allow evaluation of different hardware designs? Which of the [controllers](https://osrf.github.io/mbari_wec/architecture/#compute-and-control-systems) are the subject of ongoing software development?

MBARI also recently sponsored the development of the [LRAUV Gazebo simulator](https://github.com/osrf/lrauv). Is there a concerted effort within MBARI to focus on marine simulation capabilities?

**Comment by hamilton8415** [comment]
Yes, absolutely, thanks for the questions.

There has been considerable research in the wave-energy field to evaluate control algorithms that maximize energy capture from a given device, and often these techniques can also reduce peak loads in the system, or observe other constraints. Most of this work has been done in simulation, with considerably less testing done in real-world installations. This is understandable because developing and deploying a wave-energy converter in the ocean is a significant and costly effort. The motivation for developing this software is ultimately to provide an opportunity for researchers outside of our institute to test their algorithms on the MBARI WEC during our ongoing deployments. The simulation environment presented here is the first step towards that and provides a testing environment for any code developed. Subsequent deployment on the actual buoy should be straightforward as the buoy computers are running ROS 2 and implement the same messages and interfaces as the simulator. That step would be done with MBARI involvement after reviewing the behavior of any code in the simulation environment.  We do expect that in practice, users may need certain additional features in the simulator, our funding for this project includes the ability to provide those when possible. We also believe this simulation environment has the potential to be used in an educational environment to provide an interesting classroom experience related to the field of marine hydrodynamics, wave-energy, and controls research in general.

MBARI has developed the LRAUV simulator for similar purposes, although in that case the goal is to provide an environment to develop and test AUV missions, that again can subsequently be run on an [LRAUV ](https://www.mbari.org/technology/long-range-auv-lrauv/)at sea. Development of that simulator was done to meet our own internal needs, as well as for collaborators who plan to use, or are using, the LRAUV for scientific missions.

Within MBARI engineering, simulation work is a tool in the pursuit of applying marine robotics for oceanographic purposes, more than an activity of developing simulation capabilities as an end goal. In this case, these two projects are examples that have spilled into the open-source world and we have been very pleased to contribute the capability to correctly model added-mass effects in Gazebo. This is important for sub-sea work and we have not seen this capability in many other open-source tools.
